src/cargo/util/auth.rs

```diff
@@ -0,0 +1,444 @@
+    Ok(RegistryCredentialConfig::None)
+}
+
+/// An authorization error from accessing a registry.
+#[derive(Debug)]
+pub struct AuthorizationError {
+    /// Url that was attempted
+    pub sid: SourceId,
+    /// Url where the user could log in.
+    pub login_url: Option<Url>,
+    /// Specific message indicating what failed, such as "no token"
+    pub message: &'static str,
+}
+impl Error for AuthorizationError {}
+impl fmt::Display for AuthorizationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.sid.is_default_registry() {
+            write!(f, "{}, please run `cargo login`", self.message)
+        } else if let Some(name) = self.sid.cfg_name() {
+            write!(
+                f,
+                "{} for `{}`, please run `cargo login --registry {}`",
+                self.message,
+                self.sid.display_registry_name(),
+                name
+            )
+        } else {
+            write!(
+                f,
+                r#"{} for `{}`
+consider setting up an alternate registry in Cargo's configuration
+as described by https://doc.rust-lang.org/cargo/reference/registries.html
+
+[registries]
+my-registry = {{ index = "{}" }}
+"#,
+                self.message,
+                self.sid.display_registry_name(),
+                self.sid.url()
+            )
+        }
+    }
+}
+
+// Store a token in the cache for future calls.
+pub fn cache_token(config: &Config, sid: &SourceId, token: &str) {
+    let url = sid.canonical_url();
+    config
+        .credential_cache()
+        .insert(url.clone(), token.to_string());
+}
+
+/// Returns the token to use for the given registry.
+/// If a command_line_token is present, it will be returned and
+/// cached for future calls using the same SourceId.
+pub fn auth_token(config: &Config, sid: &SourceId, login_url: Option<&Url>) -> CargoResult<String> {
+    match auth_token_optional(config, sid)? {
+        Some(token) => Ok(token),
+        None => Err(AuthorizationError {
+            sid: sid.clone(),
+            login_url: login_url.cloned(),
+            message: "no token found",
+        }
+        .into()),
+    }
+}
+
+/// Returns the token to use for the given registry.
+fn auth_token_optional(config: &Config, sid: &SourceId) -> CargoResult<Option<String>> {
+    let mut cache = config.credential_cache();
+    let url = sid.canonical_url();
+
+    if let Some(token) = cache.get(url) {
+        return Ok(Some(token.clone()));
+    }
+
+    let credential = registry_credential_config(config, sid)?;
+    let token = match credential {
+        RegistryCredentialConfig::None => return Ok(None),
+        RegistryCredentialConfig::Token(config_token) => config_token.to_string(),
+        RegistryCredentialConfig::Process(process) => {
+            run_command(config, &process, sid, Action::Get)?.unwrap()
+        }
+    };
+
+    cache.insert(url.clone(), token.clone());
+    Ok(Some(token))
+}
+
+enum Action {
```

Comment:
From this point down in the file, it's based on existing code moved from `ops/registry/auth.rs`